br/pkg/lightning/backend/local/local.go

```diff
@@ -1342,65 +1367,46 @@ func (local *local) ImportEngine(ctx context.Context, engineUUID uuid.UUID, regi
 		}
 	}
 
-	if lf.Duplicates.Load() > 0 {
-		if err := lf.saveEngineMeta(); err != nil {
-			log.L().Error("failed to save engine meta", log.ShortError(err))
-			return err
-		}
-		log.L().Warn("duplicate detected during import engine", zap.Stringer("uuid", engineUUID),
-			zap.Int64("size", lfTotalSize), zap.Int64("kvs", lfLength), zap.Int64("duplicate-kvs", lf.Duplicates.Load()),
-			zap.Int64("importedSize", lf.importedKVSize.Load()), zap.Int64("importedCount", lf.importedKVCount.Load()))
-	}
-
 	log.L().Info("import engine success", zap.Stringer("uuid", engineUUID),
 		zap.Int64("size", lfTotalSize), zap.Int64("kvs", lfLength),
 		zap.Int64("importedSize", lf.importedKVSize.Load()), zap.Int64("importedCount", lf.importedKVCount.Load()))
 	return nil
 }
 
 func (local *local) CollectLocalDuplicateRows(ctx context.Context, tbl table.Table, tableName string, opts *kv.SessionOptions) (hasDupe bool, err error) {
-	if local.duplicateDB == nil {
-		return false, nil
-	}
-
-	logger := log.With(zap.String("table", tableName)).Begin(zap.InfoLevel, "[detect-dupe] collect duplicate local keys")
+	logger := log.With(zap.String("table", tableName)).Begin(zap.InfoLevel, "[detect-dupe] collect local duplicate keys")
 	defer func() {
 		logger.End(zap.ErrorLevel, err)
 	}()
 
-	physicalTS, logicalTS, err := local.pdCtl.GetPDClient().GetTS(ctx)
+	atomicHasDupe := atomic.NewBool(false)
+	duplicateManager, err := NewDuplicateManager(tbl, tableName, local.splitCli, local.tikvCli,
+		local.errorMgr, opts, local.dupeConcurrency, atomicHasDupe)
 	if err != nil {
-		return false, err
+		return false, errors.Trace(err)
 	}
-	ts := oracle.ComposeTS(physicalTS, logicalTS)
-	duplicateManager, err := NewDuplicateManager(local, ts, opts)
-	if err != nil {
-		return false, errors.Annotate(err, "open duplicatemanager failed")
+	if err := duplicateManager.CollectDuplicateRowsFromDupDB(ctx, local.duplicateDB, local.keyAdapter); err != nil {
```

Comment:
I prefer to return a bool that represent whether there are duplicate kvs instead of use a atomic variable parameter though the current logic is correct.

Reply:
Actually I think we should check if there are any duplicate kvs by selecting from errorMgr since `CollectDuplicateRowsFromDupDB` is not idempotent. But that would increase the changed lines of this PR, so I plan to keep using atomic variable as before and doing this from another PR.